Sygus print callbacks

ajreynol opened the pull request:
This is the first step towards addressing #1344, and also is work towards #1235.

This pull request:
- adds a generic callback mechanism for sygus datatype constructors. 
- adds a printing method in src/printer/printer.h for printing the terms that sygus datatype terms encode.

This allows much more flexibility in the future for handling artifacts from the sygus input format, and will allow us to eliminate a number of hard-coded fields in DatatypeConstructor, such as d_let_body (which is now a part of the printing callback, which can be specified as an instance of a virtual base class).

Notice that the new code in this commit has been copied from existing places in the code, such as TermDbSygus.  This code will be eliminated in a separate PR once all the components for the new printing architecture are ready (which will require 1-2 more intermediate PRs).

HanielB commented on src/expr/datatype.h:
```diff
@@ -307,6 +333,13 @@ class CVC4_PUBLIC DatatypeConstructor {
   /** is this a sygus identity function */
   bool isSygusIdFunc() const;
+  /** get sygus print callback
+   * This class stores custom ways of printing
```
The documentation is not consistent. Sometimes there is a new line after a brief, sometimes no brief and so on.

HanielB commented on src/printer/sygus_print_callback.h:
```diff
@@ -0,0 +1,109 @@
+ * This class is used for printing sygus
+ * datatype constructor terms whose top symbol
+ * is a let expression.
+ * For example, for example:
```
ditto

aniemetz commented on src/printer/smt2/smt2_printer.cpp:
```diff
@@ -1310,6 +1314,47 @@ void Smt2Printer::toStream(std::ostream& out, const Model& m, const Command* c)
+        out << dt[cIndex].getSygusOp();
+        if (n.getNumChildren() > 0)
+        {
+          for (unsigned i = 0; i < n.getNumChildren(); i++)
```
Ditto

aniemetz commented on src/printer/sygus_print_callback.cpp:
```diff
@@ -0,0 +1,130 @@
+  out << d_name;
+  if (e.getNumChildren() > 0)
+  {
+    for (unsigned i = 0; i < e.getNumChildren(); i++)
```
Ditto

ajreynol replied:
Is there a way to iterate over children of a Node with a for_each?

ajreynol commented on src/printer/smt2/smt2_printer.cpp:
```diff
@@ -364,7 +364,8 @@ void Smt2Printer::toStream(std::ostream& out, TNode n,
     out << "->";
-    for(size_t i = 0; i < n.getNumChildren(); ++i) {
+    for (size_t i = 0; i < n.getNumChildren(); ++i)
+    {
```
Mainly just doing this for consistency.  There are loops where it is useful to know the index i.

ajreynol commented:
I added a minor addition on this PR.
Any other changes needed for this one?

aniemetz commented:
Since we decided in today's meeting that we should use for each loops if possible, maybe address my corresponding comments above?

ajreynol commented:
I've addressed the review.  However, I personally don't agree that this makes the code more readable.  I think iterating over children of Node seems like there are too many implicit assumptions, such as that a Node can be iterated over, and that the semantics of the for_each returns its children.  I don't even know internally how the compiler figures this all out.  I assume it has some logic to recognize operator[] or something?

aniemetz commented:
We do have unit tests in place that check if for each works with (T)Node: https://github.com/CVC4/CVC4/pull/1230

aniemetz commented:
@ajreynol  this works because Node (and NodeValue) implements `begin()` and `end()`.

timothy-king commented:
> I assume it has some logic to recognize operator[] or something?

The answers to this are at http://en.cppreference.com/w/cpp/language/range-for . Check the "Explanation" section for the mental model/psuedo-code you probably want. As @aniemetz mentioned, this works because Node has a begin() and an end() so we fall into bullet 2 for inferring "begin_expr" and "end_expr" from "range_expression".
